src/ray/core_worker/transport/direct_actor_transport.cc

```diff
@@ -30,72 +37,78 @@ void CoreWorkerDirectActorTaskSubmitter::KillActor(const ActorID &actor_id,
-    // to have a timeout to mark it as invalid if it doesn't show up in the
-    // specified time.
-    RAY_LOG(DEBUG) << "Actor " << actor_id << " is not yet created.";
-  } else {
-    SendPendingTasks(actor_id);
-  }
+
+  SendPendingTasks(actor_id);
 }
 
 Status CoreWorkerDirectActorTaskSubmitter::SubmitTask(TaskSpecification task_spec) {
   RAY_LOG(DEBUG) << "Submitting task " << task_spec.TaskId();
   RAY_CHECK(task_spec.IsActorTask());
 
-  // We must fix the send order prior to resolving dependencies, which may complete
-  // out of order. This ensures we preserve the client-side send order.
-  int64_t send_pos = -1;
+  bool task_queued = false;
+  uint64_t send_pos = 0;
   {
     absl::MutexLock lock(&mu_);
-    send_pos = next_send_position_to_assign_[task_spec.ActorId()]++;
-  }
-
-  resolver_.ResolveDependencies(task_spec, [this, send_pos, task_spec]() mutable {
-    const auto &actor_id = task_spec.ActorId();
-
-    auto request = std::unique_ptr<rpc::PushTaskRequest>(new rpc::PushTaskRequest);
-    request->mutable_caller_address()->CopyFrom(rpc_address_);
-    // NOTE(swang): CopyFrom is needed because if we use Swap here and the task
-    // fails, then the task data will be gone when the TaskManager attempts to
-    // access the task.
-    request->mutable_task_spec()->CopyFrom(task_spec.GetMessage());
-    request->set_caller_version(caller_creation_timestamp_ms_);
-
-    absl::MutexLock lock(&mu_);
-
-    auto inserted = pending_requests_[actor_id].emplace(send_pos, std::move(request));
-    RAY_CHECK(inserted.second);
-
-    auto it = rpc_clients_.find(actor_id);
-    if (it == rpc_clients_.end()) {
-      // Actor is not yet created, or is being reconstructed, cache the request
-      // and submit after actor is alive.
-      // TODO(zhijunfu): it might be possible for a user to specify an invalid
-      // actor handle (e.g. from unpickling), in that case it might be desirable
-      // to have a timeout to mark it as invalid if it doesn't show up in the
-      // specified time.
-      RAY_LOG(DEBUG) << "Actor " << actor_id << " is not yet created.";
+    auto queue = client_queues_.find(task_spec.ActorId());
+    RAY_CHECK(queue != client_queues_.end());
+    if (queue->second.state == rpc::ActorTableData::DEAD) {
+      task_finisher_->MarkTaskCanceled(task_spec.TaskId());
+      auto status = Status::IOError("cancelling all pending tasks of dead actor");
+      // No need to increment the number of completed tasks since the actor is
+      // dead.
+      RAY_UNUSED(!task_finisher_->PendingTaskFailed(task_spec.TaskId(),
+                                                    rpc::ErrorType::ACTOR_DIED, &status));
     } else {
-      SendPendingTasks(actor_id);
+      send_pos = task_spec.ActorCounter();
+      if (send_pos < queue->second.next_send_position) {
+        // No need to resolve the dependencies again because this task has
+        // already been sent before.
```

Comment:
RAY_LOG(WARN) at least here. Unless we already log a nice message on actor task resubmission, similar to task resubmission?

Reply:
It actually goes through the same path as normal task resubmission, so it will print that "Will resubmit task..." message to the console.